### PR TITLE
[Core] Improve performance of `store` with memory pool

### LINKED ISSUE #208: [Core] Improve decoding speed while storing the KV
**Is your feature request related to a problem? Please describe.**
Currently, when using local disk backend or torch/safetensor-based serde, the vLLM's decoding throughput will be impacted when storing the KV.

**Describe the solution you'd like**
Two high-level potential techniques to fix this problem:
1. Copy out KV from GPU on a different CUDA stream
2. Avoid using the operations that may block GIL

### PR DESCRIPTION
**Pull request by YaoJiayi**
This resolves: https://github.com/LMCache/LMCache/issues/208, https://github.com/LMCache/LMCache/issues/110

Please also check the lmcache-vllm pr: https://github.com/LMCache/lmcache-vllm/pull/36

Todos:
- [x] Improve cpu save perf
- [x] Improve disk save perf
- [x] Modify config to make mem pool aware of vllm info
---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to LMCache! Before submitting the pull request, please ensure the PR meets the following criteria. This helps us maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Please try to classify PRs for easy understanding of the type of changes. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Core]</code> for changes in the core LMCache logic (e.g., <code>LMCacheEngine</code>, <code>Backend</code> etc.)</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>Pass all linter checks. Please use <a href="https://github.com/LMCache/LMCache/blob/dev/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li> Please include sufficient tests to ensure the change is stay correct and robust. This includes both unit tests and integration tests.</li>
</ul>

<h3>What to Expect for the Reviews</h3>

We aim to address all PRs in a timely manner. If no one reviews your PR within 5 days, please @-mention one of KuntaiDu, ApostaC or YaoJiayi.

### CODE REVIEW

File: lmcache/config.py [hunk 1]
```diff
@@ -16,7 +17,18 @@ class LMCacheEngineMetadata:
...
+    """ the data type of kv tensors """
+    """ (num_layer, 2, chunk_size, num_kv_head, head_size) """
+    kv_shape: tuple[int, int, int, int, int]
+
+
+@dataclass
+class LMCacheMemPoolMetadata:
+    """ Subset of `LMCacheEngineMetadata` to initialize MemPool"""
+
+    kv_shape: Tuple[int, int, int, int, int]
```
Comment (ApostaC) [inline lmcache/config.py:30]: What's the meaning of each dimension in this 5-tuple? Let's have a docstring for this.

Also, are we going to support different layouts and shapes? For example, huggingface transformers has a different layout, and seems like when running vLLM on V100 will have a different shape)

File: lmcache/storage_backend/mem_pool/local_pool.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+        """
+        self.free_pool.append(kv_obj.chunk_idx)
+
+
+class LocalCPUPool(LocalPool):
+
+    def __init__(self, metadata: LMCacheMemPoolMetadata):
+        self.chunk_size = metadata.kv_shape[2]
+        # TODO(Jiayi): the `max_chunk_num` should be computed
+        # from `config.max_cache_size`
```
Comment (ApostaC) [inline lmcache/storage_backend/mem_pool/local_pool.py:59]: Please create a new issue for this if you want to do it in the future, thanks!

File: lmcache/storage_backend/mem_pool/local_pool.py [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+
+class LocalCPUBufferPool(LocalCPUPool):
+
+    def allocate(self, kv_chunk: torch.Tensor) -> Optional[KVObj]:
+        num_tok = kv_chunk.shape[2]
+        assert num_tok <= self.chunk_size
+        while not self.free_pool:
+            logger.info("No free memory chunks. Waiting...")
+            return None
```
Comment (ApostaC) [inline lmcache/storage_backend/mem_pool/local_pool.py]: Why do we use "while" here? We will return in the first iteration anyway, so I guess "if" will be the same?

File: lmcache/storage_backend/mem_pool/local_pool.py
Comment (ApostaC) [inline lmcache/storage_backend/mem_pool/local_pool.py:1]: Btw, do we need some unit tests for "allocate" and "free"?

File: lmcache/storage_backend/local_backend.py [hunk 4]
```diff
@@ -126,31 +126,58 @@ def put_nonblocking(self, key, kv_chunk):
...
+        # Allocate the kv chunk
+        kv_obj = self.mpool.allocate(kv_chunk)
+        self.update_lock.release()
+
+        if kv_obj is None:
+            return
+
+        put_stream = torch.cuda.Stream()
+        if kv_chunk.device != torch.cpu:
+            put_stream.wait_stream(torch.cuda.default_stream(kv_chunk.device))
```
Comment (ApostaC) [inline lmcache/storage_backend/local_backend.py]: Not sure about this: do we need to `wait_stream` here, or do it at the beginning of `batched_put`?

Comment (ApostaC) [reply lmcache/storage_backend/local_backend.py]: Actually, I feel like the local CPU Backend does not need a separate thread. Let's talk about this tomorrow?

File: lmcache/storage_backend/local_backend.py [hunk 5]
```diff
@@ -311,38 +353,105 @@ def remove(
...
+    def save_disk(
+        self,
+        path: str,
+        kv_chunk: torch.Tensor,
+    ):
+        save_file({"kv_chunk": kv_chunk.contiguous()}, path)
+
+    @_lmcache_nvtx_annotate
+    @torch.inference_mode()
+    def put_nonblocking(
```
Comment (ApostaC) [inline lmcache/storage_backend/local_backend.py:377]: I like the idea of using multi-process to save the cache. In this case, I also feel like we can get rid of the thread during the put.

**Review by ApostaC — COMMENTED** [review]
Thanks, Jiayi! Let's discuss this on Monday!

File: lmcache/storage_backend/local_backend.py (hunk 4 above)
Comment (YaoJiayi) [reply lmcache/storage_backend/local_backend.py]: Updated the comment. The wait is needed as there are operations done on that chunk by the main stream. Will try to remove the extra threads.

File: lmcache/config.py (hunk 1 above)
Comment (ApostaC) [reply lmcache/config.py:30]: Will be done in #117

File: lmcache/storage_backend/local_backend.py (hunk 5 above)
Comment (YaoJiayi) [reply lmcache/storage_backend/local_backend.py:377]: Will try to solve this in https://github.com/LMCache/LMCache/issues/207

File: lmcache/storage_backend/mem_pool/base_pool.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+class BasePool(metaclass=abc.ABCMeta):
+    """
+    Interface for mem pool
+    """
+
+    @abc.abstractmethod
+    def allocate(self, kv_chunk: torch.Tensor) -> Optional[KVObj]:
+        """
+        Allocate a buffer memory pointer from the memory pool.
```
Comment (ApostaC) [inline lmcache/storage_backend/mem_pool/base_pool.py:22]: Just to double check, the `kv_chunk` is only used to check the shape, right? We won't really copy the data into the allocated chunk.

**Review by ApostaC — APPROVED** [review]
LGTM, thanks!

File: lmcache/storage_backend/mem_pool/base_pool.py (hunk 6 above)
Comment (YaoJiayi) [reply lmcache/storage_backend/mem_pool/base_pool.py:22]: No real data copying is happening here:)

**Pull request merged**